### PR TITLE
PDDL System Upgrade

### PR DESCRIPTION
**Pull request by ASzot**
## Motivation and Context

- Cleans up the PDDL system. Before we assumed the existence of some PDDL types that the user had to include in the PDDL file. These are no longer assumed.
- Added Rearrange simulator types. These are important to distinguish which objects are which in the simulator. For example, how do I know if a rigid object is a "movable object" or a "receptacle"? The physics mode of the object is not enough to tell. 
- Cleaned up rearrange simulator to not expose some private methods and variables.
- Added some doc strings.

## How Has This Been Tested

- Working in end-to-end training. 

## Types of changes

- **\[Docs change\]** Addition or changes to the documentation
- **\[Refactoring\]** Large changes to the code that improve its functionality or performance
- **\[Development\]** A pull request that add new features to the [habitat-lab](/habitat-lab) task and environment codebase.

### CODE REVIEW

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py [hunk 1]
```diff
@@ -226,15 +259,15 @@ def parse_predicate(
             ) from e
         return pred
 
-    def _parse_only_logical_expr(
+    def parse_only_logical_expr(
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py:262]: Why make this method public? It looks like it is only used in this file anyways.

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py [hunk 2]
```diff
@@ -528,22 +611,7 @@ def solution(self):
 
     @property
     def all_entities(self) -> Dict[str, PddlEntity]:
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py:613]: Can't you just inherit from the PDDLDomain class ?

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_predicate.py [hunk 3]
```diff
@@ -83,9 +83,17 @@ def sub_in(self, sub_dict: Dict[PddlEntity, PddlEntity]) -> "Predicate":
         return self
 
     def is_true(self, sim_info: PddlSimInfo) -> bool:
+        """
+        Returns if the predicate is satisfied in the current simulator state.
+        """
+
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_predicate.py:89]:
```suggestion
```

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_predicate.py [hunk 4]
```diff
@@ -83,9 +83,17 @@ def sub_in(self, sub_dict: Dict[PddlEntity, PddlEntity]) -> "Predicate":
...
         return self._pddl_sim_state.is_true(sim_info)
 
     def set_state(self, sim_info: PddlSimInfo) -> None:
+        """
+        Sets the simulator state to satisfy the predicate.
+        """
+
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_predicate.py:96]:
```suggestion
```
Nice docstring

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_sim_state.py [hunk 5]
```diff
@@ -53,15 +53,21 @@ def sample(self) -> float:
...
 
 
-@dataclass
+@dataclasses.dataclass
 class PddlRobotState:
     """
-    Specifies the configuration of the robot. Only used as a data structure. Not used to set the simulator state.
+    Specifies the configuration of the robot.
+
+    :property place_at_pos_thresh: If -1.0, this will place the robot as close
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_sim_state.py:61]: Docstring for base_angle_noise

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_sim_state.py [hunk 6]
```diff
@@ -305,15 +354,34 @@ def set_state(self, sim_info: PddlSimInfo) -> None:
         """
         sim = sim_info.sim
         for entity, target in self._obj_states.items():
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_sim_state.py:356]: Can you add a comment on what this loop does ?

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py [hunk 7]
```diff
@@ -446,9 +495,43 @@ def get_possible_actions(
                     matching_actions.append(new_action)
         return matching_actions
 
+    def get_ordered_actions(self) -> List[PddlAction]:
+        """
+        Gets an ordered list of PDDL actions.
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py:500]: which actions? I assume the ones that are available to the agent as opposed to the ones that have been used so far, etc... 
"an ordered list of PDDL actions" is not descriptive

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py [hunk 8]
```diff
@@ -446,9 +495,43 @@ def get_possible_actions(
...
+    def get_entity(self, k: str) -> PddlEntity:
+        """
+        Gets an entity from the `all_entities` dictionary by key name.
+        """
+
+        return self.all_entities[k]
+
+    def find_entities(self, entity_type: ExprType) -> Iterable[PddlEntity]:
+        """
+        Returns the all entities that match the condition.
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py:516]:
```suggestion
        Returns all the entities that match the condition.
```

**Review by vincentpierre — APPROVED** [review]
I think this is an improvement, but I am still unsure how the code works.

**Review by rpartsey — APPROVED** [review]

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py (hunk 1 above)
Comment (ASzot) [reply habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py:255]: This is an important API to expose to code outside of Habitat-Lab / Habitat-Baselines for setting up custom tasks.

File: habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py (hunk 2 above)
Comment (ASzot) [reply habitat-lab/habitat/tasks/rearrange/multi_task/pddl_domain.py:643]: What do you mean? The PDDLDomain class provides some base entities and then PDDLProblem adds more entities to this dict.

**Pull request merged**